lib/WUI/wui_REST_api.c

```diff
@@ -12,82 +12,169 @@
+        "\"target\": 0,"
+        "\"offset\": 0"
+        "},"
+        "\"bed\": {"
+        "\"actual\": %d.%.1d,"
+        "\"target\": 0,"
+        "\"offset\": 0"
+        "},"
+        "\"chamber\": {"
+        "\"actual\": 0,"
+        "\"target\": 0,"
+        "\"offset\": 0"
+        "}"
+        "},"
+        "\"sd\": {"
+        "\"ready\": 1"
+        "},"
+        "\"state\": {"
+        "\"text\": \"Operational\","
+        "\"flags\": {"
+        "\"operational\": %ld,"
+        "\"paused\": %ld,"
+        "\"printing\": %ld,"
+        "\"cancelling\": %ld,"
+        "\"pausing\": %ld,"
+        "\"sdReady\": %ld,"
+        "\"error\": %ld,"
+        "\"ready\": %ld,"
+        "\"closedOrError\": %ld,"
+        "\"busy\": %ld"
+        "}"
+        "}"
+        "}",
+        filament_material,
+        (int)wui_vars_copy.temp_nozzle, (int)((wui_vars_copy.temp_nozzle - (int)wui_vars_copy.temp_nozzle) * 10),
+        (int)wui_vars_copy.temp_bed, (int)((wui_vars_copy.temp_bed - (int)wui_vars_copy.temp_bed) * 10),
+        operational, paused, printing, cancelling, pausing, sd_ready,
+        error, ready, closed_on_error, busy);
+}
 
-    if (wui_vars_copy.time_to_end == TIME_TO_END_INVALID) {
-        time_to_end = 0;
-    }
+void get_version(char *data, const uint32_t buf_len) {
+    snprintf(data, buf_len,
+        "{"
+        "\"api\": \"0.1\","
+        "\"server\": \"2.0.0\","
+        "\"text\": \"Prusa Local MINI 2.0.0\","
+        "\"hostname\": \"prusa-mini\""
+        "}");
+}
 
-    print_dur_to_string(print_time, sizeof(print_time), wui_vars_copy.print_dur);
+void get_job(char *data, const uint32_t buf_len) {
+
+    osStatus status = osMutexWait(wui_thread_mutex_id, osWaitForever);
+    if (status == osOK) {
+        wui_vars_copy = wui_vars;
+    }
+    osMutexRelease(wui_thread_mutex_id);
 
-    snprintf(data, buf_len, "{"
-                            "\"temp_nozzle\":%ld,"
-                            "\"temp_bed\":%ld,"
-                            "\"material\":\"%s\","
-                            "\"pos_z_mm\":%.2f,"
-                            "\"printing_speed\":%d,"
-                            "\"flow_factor\":%d,"
-                            "\"progress\":%d,"
-                            "\"print_dur\":\"%s\","
-                            "\"time_est\":\"%lu\","
-                            "\"time_zone\":\"%d\","
-                            "\"project_name\":\"%s\""
-                            "}",
-        actual_nozzle, actual_heatbed, filament_material,
-        z_pos_mm, print_speed, flow_factor, wui_vars_copy.sd_precent_done,
-        print_time, time_to_end, time_zone, wui_vars_copy.gcode_name);
+    snprintf(data, buf_len,
+        "{"
+        "\"job\":{"
+        "\"estimatedPrintTime\":%ld,"
+        "\"file\":{"
+        "\"date\":null,"
+        "\"name\":\"%s\","
+        "\"origin\":\"USB\","
+        "\"path\":\"%s\","
+        "\"size\":%ld"
+        "}"
+        "},"
```

Comment:
/suggestion/ Use proper tool for json serialization. It is only a matter of time until this introduces a nasty bug/security issue.

It might be actually very easy – the only issue I am aware of is string serialization (`%s`), where we need to escape `"` characters. As we have our own `[vsn]printf()` implementation (see `printf.c`), we can easily add a new one implementing the escaping (something line `jsonprintf(...)`).

Reply:
Implement function as you proposed is not part of this PR. There must be specified what we want/need from such a method and at least set of unit tests for it. It should be separate task and unfortunately there isn't any method to use to compose json response.